pkg/apps/application-list.jsx

```diff
@@ -0,0 +1,168 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import React from "react";
+
+import PackageKit from "./packagekit";
+import { left_click, icon_url, show_error, launch, ProgressBar, CancelButton } from  "./utils.jsx";
+
+const _ = cockpit.gettext;
+
+class ApplicationRow extends React.Component {
+    constructor() {
+        super();
+        this.state = { progress: false };
+    }
+
+    render() {
+        var self = this;
+        var comp = self.props.comp;
+        var state = self.state;
+
+        function navigate() {
+            cockpit.location.go(comp.id);
+        }
+
+        function action(func, arg, progress_title) {
+            self.setState({ progress_title: progress_title });
+            func(arg, (data) => self.setState({ progress: data })).
+                always(() => self.setState({ progress: false })).
+                fail(show_error);
+        }
+
+        function install() {
+            action(PackageKit.install, comp.pkgname, _("Installing"));
+        }
+
+        function remove() {
+            action(PackageKit.remove, comp.file, _("Removing"));
+        }
+
+        var name, summary_or_progress, button;
+
+        if (comp.installed) {
+            name = <a onClick={left_click(() => launch(comp))}>{comp.name}</a>;
+        } else {
+            name = comp.name;
+        }
+
+        if (state.progress) {
```

Comment:
Isn't 0% a valid progress too? That's why I'd rather prefer a more explicit default value, or using -1 and checking for `>= 0` here. Also, PolicyKit occasionally uses 101% as a value for "I can't determine this" (see `percentage` on https://www.freedesktop.org/software/PackageKit/gtk-doc/Transaction.html)

Reply:
> Isn't 0% a valid progress too? 

Yes, good point.

> Also, PolicyKit occasionally uses 101% 

This is filtered out in packagekit.js